File: scenarios/command/dotnet/proto_messages/ProtobufSerializer.cs
﻿
using Google.Protobuf;
using MQTTnet.Client.Extensions;

namespace proto_messages;

public class ProtobufSerializer : IMessageSerializer
{
    private readonly MessageParser? _parser;

    public ProtobufSerializer(MessageParser parser) => _parser = parser;

    public string ContentType => "application/protobuf";

    public T FromBytes<T>(byte[] payload) => (T)_parser!.ParseFrom(payload);

    public byte[] ToBytes<T>(T payload) => (payload as IMessage).ToByteArray();

}

File: scenarios/command/ts/commandClient/src/protobufSerializer.ts

import {
    IMessageSerializer
} from '@mqttapplicationsamples/mqttjsclientextensions';
import {
    IUnlockRequest,
    UnlockRequest
} from '@mqttapplicationsamples/protomessages';

export class ProtobufSerializer implements IMessageSerializer {
    public contentType = "application/protobuf";

    public fromBytes<T>(payload: Buffer): T {
        return UnlockRequest.decode(payload) as T;
    }

    public toBytes<T>(payload: T): Buffer {
        return Buffer.from(UnlockRequest.encode(payload as IUnlockRequest).finish());
    }
}
